powershell cli video post (#1502)
oktadeveloper/okta-blog: _source/_posts/2024-04-11-okta-powershell-module-podcast.md

---
layout: blog_post
title: "Introducing Okta's Official PowerShell Module"
author: edunham
by: advocate
communities: [devops,.net]

> description: "In this podcast, we discuss the new PowerShell module with Laura Rodriguez and Bhavik Thakkar"

Reviewer: ```suggestion
description: "In this podcast, we discuss the new PowerShell module that enables you to manage your Okta resources with Laura Rodriguez and Bhavik Thakkar"
```

Recommending adding a "why is this something one would watch" in here since this is the description that shows as a short preview for the link (and within Google search results). But please feel free to change the text to whatever is appropriate

> tags: [powershell-cli]

Reviewer: IMO i recommend this be split out into 2 tags, `powershell` and `cli`

> tags: [powershell-cli]

Author: ```suggestion
tags: [powershell,cli]
```

tweets:
- ""
- ""
- ""
image: blog/identity-developer-podcast/podcast.jpg
type: awareness
---

The Okta Workforce Identity Developer Podcast returns with an exploration of our latest new developer tool, the Okta PowerShell Module! 

> ## Watch on YouTube

Reviewer: ```suggestion
## Watch on the OktaDev YouTube channel
```

{% youtube ADnxvCYLL9A %}

You can find the source of the Okta PowerShell Module, and documentation in the README, [on GitHub](https://github.com/okta/okta-powershell-cli). 

Below is the example code discussed in the podcast episode.

## Example: Interactively create Okta groups and group rules with PowerShell

### Install PowerShell module via PSGallery

```ps
Install-Module -Name Okta.PowerShell
```

> ### Set up configuration

Reviewer: ```suggestion
### Set up the OAuth 2.0 configuration
```

(trying to work in 0Auth into a header)

Set up an app integration following [this guide](https://developer.okta.com/docs/guides/device-authorization-grant/main/#configure-an-application-to-use-the-device-authorization-grant) to get the Client ID. Then configure PowerShell: 

```ps
$Configuration = Get-OktaConfiguration
$Configuration.BaseUrl = $env:OKTA_PS_ORG_URL
$Configuration.ClientId = $env:OKTA_PS_CLIENT_ID
$Configuration.Scope = "okta.groups.manage okta.apps.manage okta.users.manage"
```

### Get an access token

Running this command in PowerShell opens a browser so you can follow your organization's usual login flow. Logging into Okta in your browser provides a token to the PowerShell session that opened the browser. 

```ps
Invoke-OktaEstablishAccessToken
```

> ### Group Creation

Reviewer 2: ```suggestion
### Group creation
```

> ### Group creation

Reviewer: ```suggestion
### Create an Okta group using PowerShell
```

(finding more opportunities to work in keywords)

```ps
$GroupProfile = [PSCustomObject]@{
                name = "Sales Team"
                description = "All employees that belong to the Sales team"
            }
$NewGroup = Initialize-OktaGroup -VarProfile $GroupProfile

$CreatedGroup = New-OktaGroup -Group $NewGroup

echo $CreatedGroup
```

> ### Group Rules Definition

Reviewer 2: ```suggestion
### Group rules definition
```

```ps

# For example, a group rule may specify that users with the job title "Sales Representative" are automatically added to the "Sales Team" group.

# List all users which title is "Sales Representative"
Invoke-OktaListUsers -Search 'profile.title eq "Sales Representative"'

$NewGroupRule = @{
    name = "Assign users to the Sales Team"
    type = "group_rule"
    actions = @{
        assignUserToGroups = @{
            groupIds = @($CreatedGroup.Id)
        }
    }  
    conditions = @{
        expression = @{
            type = "urn:okta:expression:1.0"
            value = "user.title=='Sales Representative'"
        }
    }
}

$CreatedRule = New-OktaGroupRule -GroupRule $NewGroupRule -IncludeNullValues

Echo $CreatedRule

Invoke-OktaActivateGroupRule -RuleId $CreatedRule.Id    

Get-OktaGroupRule -RuleId $CreatedRule.Id

Invoke-OktaListGroupUsers -GroupId $CreatedGroup.Id
```

> ### Onboarding New Employees

Reviewer 2: ```suggestion
### Onboarding new employees
```

```ps
$UserProfile = [PSCustomObject]@{
    firstName     = 'Lionel'
    lastName = 'Messi'
    login = 'lio.messi@mailinator.com'
    email = 'lio.messi@mailinator.com'
    title = 'Sales Representative'
    }

$CreateUserRequest = Initialize-OktaCreateUserRequest -VarProfile $UserProfile

New-OktaUser -Body $CreateUserRequest

Invoke-OktaListGroupUsers -GroupId $CreatedGroup.Id
```

> ## Example: Script to automatically sync employee data from a file to Okta

Reviewer 2: ```suggestion
## Example: script to automatically sync employee data from a file to Okta
```

> For this scenario, we imagine that a custom tool which lacks OIDC or SCIM compatibility has emitted a CSV of users who we'll add to our Okta organization using a PowerShell script. 

Reviewer 2: ```suggestion
For this scenario, we imagine a custom tool that lacks [OIDC](https://developer.okta.com/blog/2023/07/28/oidc_workshop) or [SCIM](https://developer.okta.com/blog/2023/07/28/scim-workshop) compatibility has emitted a CSV of users who we'll add to our Okta organization using a PowerShell script. 
```

> For this scenario, we imagine a custom tool that lacks [OIDC](https://developer.okta.com/blog/2023/07/28/oidc_workshop) or [SCIM](https://developer.okta.com/blog/2023/07/28/scim-workshop) compatibility has emitted a CSV of users who we'll add to our Okta organization using a PowerShell script. 

Reviewer: ```suggestion
For this scenario, we imagine a custom tool that lacks [OpenID Connect (OIDC)](https://developer.okta.com/blog/2023/07/28/oidc_workshop) or [System for Cross-Identity Management (SCIM)](https://developer.okta.com/blog/2023/07/28/scim-workshop) compatibility has emitted a CSV of users who we'll add to our Okta organization using a PowerShell script. 
```

...
``` 

The following PowerShell script creates users in Okta based on the CSV data: 

```ps
$filePath = Resolve-Path ~/Documents/hr-export-employees.csv
if (Test-Path $filePath) {
    # Read the CSV file
    $tableData = Import-Csv $filePath
    # Iterate over each row
    foreach ($row in $tableData) {
        $userProfile = [PSCustomObject]@{}
        # Iterate over each property in the row
        foreach ($property in $row.PSObject.Properties) {
            Add-Member -InputObject $userProfile -MemberType NoteProperty -Name $property.Name -Value $property.Value
            if ($property.Name -eq "email") {
                Add-Member -InputObject $userProfile -MemberType NoteProperty -Name "login" -Value $property.Value
            }
            Write-Host "$($property.Name): $($property.Value)"

        }
        Write-Host  $userProfile
        $createUserRequest = Initialize-OktaCreateUserRequest -VarProfile $userProfile
        $createdUser = New-OktaUser -Body $CreateUserRequest
        Write-Host $createdUser
    }  
 }

    Write-Host "-----------------------------------------"
} else {
    Write-Host "File not found: $filePath"
}
```

> What will you automate with Okta's PowerShell module? Let us know in the comments below!

Reviewer: Should add the Follow us CTA

> What will you automate with Okta's PowerShell module? Let us know in the comments below!

Author: ```suggestion
What will you automate with Okta's PowerShell module? Let us know in the comments below!  

Follow us on [Twitter](https://twitter.com/oktadev) and subscribe to our [YouTube](https://www.youtube.com/c/oktadev) channel. If you have any questions or you want to share what other topics you'd like to hear about on the podcast, please comment below!
```

---
layout: blog_post
title: "Introducing Okta's Official PowerShell Module"
author: edunham
by: advocate
communities: [devops,.net]
description: "In this podcast, we discuss the new PowerShell module that enables you to manage your Okta resources with Laura Rodriguez and Bhavik Thakkar"
tags: [powershell,cli]
tweets:
- ""
- ""
- ""
image: blog/identity-developer-podcast/podcast.jpg
type: awareness
---

The Okta Workforce Identity Developer Podcast returns with an exploration of our latest new developer tool, the Okta PowerShell Module! 

## Watch on the OktaDev YouTube channel

{% youtube ADnxvCYLL9A %}

You can find the source of the Okta PowerShell Module, and documentation in the README, [on GitHub](https://github.com/okta/okta-powershell-cli). 

Below is the example code discussed in the podcast episode.

## Example: Interactively create Okta groups and group rules with PowerShell

### Install PowerShell module via PSGallery

```ps
Install-Module -Name Okta.PowerShell
```

### Set up the OAuth 2.0 configuration

> Set up an app integration following [this guide](https://developer.okta.com/docs/guides/device-authorization-grant/main/#configure-an-application-to-use-the-device-authorization-grant) to get the Client ID. Then configure PowerShell: 

Author: ```suggestion
Set up an app integration following [this guide](/docs/guides/device-authorization-grant/main/#configure-an-application-to-use-the-device-authorization-grant) to get the Client ID. Then configure PowerShell: 
```

Invoke-OktaListUsers -Search 'profile.title eq "Sales Representative"'

$NewGroupRule = @{
    name = "Assign users to the Sales Team"
    type = "group_rule"
    actions = @{
        assignUserToGroups = @{
            groupIds = @($CreatedGroup.Id)
        }
    }  
    conditions = @{
        expression = @{
            type = "urn:okta:expression:1.0"
            value = "user.title=='Sales Representative'"
        }
    }
}

$CreatedRule = New-OktaGroupRule -GroupRule $NewGroupRule -IncludeNullValues

Echo $CreatedRule

Invoke-OktaActivateGroupRule -RuleId $CreatedRule.Id    

Get-OktaGroupRule -RuleId $CreatedRule.Id

Invoke-OktaListGroupUsers -GroupId $CreatedGroup.Id
```

### Onboarding new employees

```ps
$UserProfile = [PSCustomObject]@{
    firstName     = 'Lionel'
    lastName = 'Messi'
    login = 'lio.messi@mailinator.com'
    email = 'lio.messi@mailinator.com'
    title = 'Sales Representative'
    }

$CreateUserRequest = Initialize-OktaCreateUserRequest -VarProfile $UserProfile

New-OktaUser -Body $CreateUserRequest

Invoke-OktaListGroupUsers -GroupId $CreatedGroup.Id
```

## Example: script to automatically sync employee data from a file to Okta

> For this scenario, we imagine a custom tool that lacks [OpenID Connect (OIDC)](https://developer.okta.com/blog/2023/07/28/oidc_workshop) or [System for Cross-Identity Management (SCIM)](https://developer.okta.com/blog/2023/07/28/scim-workshop) compatibility has emitted a CSV of users who we'll add to our Okta organization using a PowerShell script. 

Author: ```suggestion
For this scenario, we imagine a custom tool that lacks [OpenID Connect (OIDC)](/blog/2023/07/28/oidc_workshop) or [System for Cross-Identity Management (SCIM)](/blog/2023/07/28/scim-workshop) compatibility has emitted a CSV of users who we'll add to our Okta organization using a PowerShell script. 
```
